hpx/lcos/local/detail/condition_variable.hpp

```diff
@@ -115,20 +120,36 @@ namespace hpx { namespace lcos { namespace local { namespace detail
                 queue_.front().id_ = threads::invalid_thread_id_repr;
                 queue_.pop_front();
 
-                util::scoped_unlock<Lock> unlock(lock);
+                if (!queue_.empty())
+                {
+                    util::scoped_unlock<Lock> unlock(lock);
 
-                threads::set_thread_state(threads::thread_id_type(
-                    reinterpret_cast<threads::thread_data_base*>(id)),
-                    threads::pending, threads::wait_timeout,
-                    threads::thread_priority_default, ec);
-                if (!ec) return true;
+                    threads::set_thread_state(threads::thread_id_type(
+                        reinterpret_cast<threads::thread_data_base*>(id)),
+                        threads::pending, threads::wait_timeout,
+                        threads::thread_priority_default, ec);
+                    if (!ec) return true;
+                }
+                else
+                {
+                    // Since this is the last thread waiting on this condition
+                    // variable it could happen that this instance will be
+                    // destroyed before set_thread_state() returns. We have to
+                    // make sure that this instance is not touched anymore.
+                    lock.unlock();
```

Comment:
I went through all current uses of `detail::condition_variable::notify_one` and fixed its uses. Also a unlocked lock will not touch the mutex anymore (at least not `boost::unique_lock` which we usually use as the scoped lock mechanism). All of this is not too satisfying, but I believe it's the best we can currently do.